From e97f679bc571391d250b500d973ff6843623dcbf Mon Sep 17 00:00:00 2001
From: Hari Naraayana Desikan Kannan <quic_hnaraaya@quicinc.com>
Date: Tue, 10 Sep 2024 13:15:07 +0530
Subject: [PATCH] Trigger scan when bss load update is enabled

Currently, when bss_load_update_period is configured with a non-zero value,
the survey list is fetched. When no scan is triggered, this leads to
scan_link being NULL, causing a hostapd crash. To address this issue,
trigger a scan in update_channel_utilization() when survey list is empty
and skip scan result fetch if scan_link is NULL.

Signed-off-by: Hari Naraayana Desikan Kannan <quic_hnaraaya@quicinc.com>
---
 src/ap/bss_load.c            | 24 ++++++++++++++++++++++++
 src/drivers/driver_nl80211.c | 11 +++++++++++
 2 files changed, 35 insertions(+)

--- a/src/ap/bss_load.c
+++ b/src/ap/bss_load.c
@@ -45,6 +45,7 @@ static void update_channel_utilization(v
 	unsigned int sec, usec;
 	int err;
 	struct hostapd_iface *iface = hapd->iface;
+	struct wpa_driver_scan_params params;
 
 	if (!(hapd->beacon_set_done && hapd->started))
 		return;
@@ -53,7 +54,26 @@ static void update_channel_utilization(v
 				     hapd->iface->conf->acs_exclude_6ghz_non_psc);
 	if (err) {
 		wpa_printf(MSG_ERROR, "BSS Load: Failed to get survey data");
-		return;
+		os_memset(&params, 0, sizeof(params));
+		params.freqs = os_calloc(2, sizeof(params.freqs[0]));
+
+		if(params.freqs == NULL)
+			return;
+
+		params.freqs[0] = hapd->iface->freq;
+		if (hapd->iface->num_multi_hws) {
+			params.bssid = hapd->iface->bss[0]->conf->bssid;
+			wpa_printf(MSG_DEBUG, "scan triggered with bssid" MACSTR "\n",
+				   MAC2STR(params.bssid));
+		}
+
+		err = hostapd_driver_scan(hapd, &params);
+		if (err < 0) {
+			wpa_printf(MSG_DEBUG,
+				   "Failed to request a scan ret=%d (%s) in bss_load_update_init",
+				   err, strerror(-err));
+		}
+		os_free(params.freqs);
 	}
 
 	ieee802_11_set_beacon(hapd);
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -10523,6 +10523,11 @@ static int wpa_driver_nl80211_get_survey
 					 &arg, NULL, NULL);
 	} while (err > 0);
 
+	if (!bss->scan_link) {
+		wpa_printf(MSG_ERROR, "nl80211: scan_link empty");
+		return -1;
+	}
+
 	if (err)
 		wpa_printf(MSG_ERROR, "nl80211: Failed to process survey data");
 	else
